File: Competitive Programming 3/Section 1.4/Anagram/156 - Ananagrams/Python/main.py
from sys import stdin

sortedLowerStrings = {}
anagramOriginal = []

def main():
    for line in stdin:
        line = line.rstrip()
        if line[0]=='#':
            break
        for word in line.split(' '):
            if word:
                anagramOriginal.append([word,False])
                sortedLowerWord = list(word.lower())
                sortedLowerWord.sort()
                sortedLowerWord = ''.join(sortedLowerWord)
                if sortedLowerWord in sortedLowerStrings :
                    #visit both
                    anagramOriginal[sortedLowerStrings[sortedLowerWord]][1]=anagramOriginal[len(anagramOriginal)-1][1]=True
                else:
                    sortedLowerStrings[sortedLowerWord]=len(anagramOriginal)-1
    anagramOriginal.sort()
    for originalP in anagramOriginal:
        if not originalP[1]:
            print(originalP[0])
main()

File: Competitive Programming 3/Section 1.4/Anagram/156 - Ananagrams/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

map<string,int> sortedLowerStrings;
vector<pair<string,bool> > originalAnagram;

void processWord(string & word){
    originalAnagram.push_back(make_pair(word,false));
    transform(word.begin(),word.end(),word.begin(),[](char c){
        return tolower(c);
    });
    sort(word.begin(),word.end());
    auto it = sortedLowerStrings.find(word);
    if(it == sortedLowerStrings.end())
        sortedLowerStrings[word]=originalAnagram.size()-1;
    else
        //visit both
        originalAnagram[it->second].second=originalAnagram[originalAnagram.size()-1].second=true;
}
int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    string word;
    while(cin>>word&&word[0]!='#'){
        processWord(word);
    }
    sort(originalAnagram.begin(),originalAnagram.end());
    for(auto oA : originalAnagram)
        if(!oA.second)
            cout << oA.first << "\n";
    return 0;
}